game_events.go

```diff
@@ -564,18 +568,66 @@ func (geh gameEventHandler) nadeEvent(data map[string]*msg.CSVCMsg_GameEventKeyT
 
 	return events.GrenadeEvent{
 		GrenadeType:     nadeType,
+		Grenade:         geh.getThrownGrenade(thrower, nadeType),
 		Thrower:         thrower,
 		Position:        position,
 		GrenadeEntityID: nadeEntityID,
 	}
 }
 
-func mapGameEventData(d *msg.CSVCMsg_GameEventListDescriptorT, e *msg.CSVCMsg_GameEvent) map[string]*msg.CSVCMsg_GameEventKeyT {
-	data := make(map[string]*msg.CSVCMsg_GameEventKeyT)
-	for i, k := range d.Keys {
-		data[k.Name] = e.Keys[i]
-	}
-	return data
+func (geh gameEventHandler) addThrownGrenade(p *common.Player, wep *common.Equipment) {
+
+    if p != nil {
+
+        gameState := geh.gameState()
+        gameState.thrownGrenades[p] = append(gameState.thrownGrenades[p], wep)
+    }
+}
+
+func (geh gameEventHandler) getThrownGrenade(p *common.Player, wepType common.EquipmentElement) *common.Equipment {
+
+    if p != nil {
+        // Get the first weapon we found for this player with this weapon type
+        for _, thrownGrenade := range geh.gameState().thrownGrenades[p] {
+            if(thrownGrenade.Weapon == wepType) {
+                return thrownGrenade
+            }
+        }
+    }
+
+	// If we didn't found the thrown grenade we send back a new Weapon of the correct type (so we don't break anything)
+	thrownGrenade := common.NewEquipment(wepType)
+	return &thrownGrenade
```

Comment:
I think we should return nil here.

Otherwise it might be very confusing why there is a `*Equipment` with a new `UniqueID()` that was never there before.
It could also suppress bugs which we shouldn't do.

If it doesn't occur then great, otherwise we can document it on the events.

Reply:
After thinking well you are probably right!

That would means that the following values might be nil for not found thrown grenades (which hopefully shouldn't happen ...): 

- events.GrenadeEvent.Grenade
- events.Kill.Weapon
- events.PlayerHurt.Weapon


Are you still ok with that?

PS: I commited some last changes that you should review, but after that I think we should be good to go!

Reply:
ahhh shit, I didn't think that through.

`Kill.Weapon` and `PlayerHurt.Weapon` shouldn't be nil since people are already using those.

I will test if this happens in any of the regression set. if it is never nil I think it's ok to do it like this as `nil` would then just indicate a bug.